src/modules/dwm.cpp

```diff
@@ -0,0 +1,544 @@
+        auto tags = m_ipc->get_tags();
+        for (dwmipc::Tag& t : *tags) {
+          auto state = get_state(t.bit_mask);
+          auto label = m_state_labels.at(state)->clone();
+          label->replace_token("%name%", t.tag_name);
+          m_tags.emplace_back(t.tag_name, t.bit_mask, state, move(label));
+        }
+
+        // This event is for keeping track of the tag states
+        m_ipc->on_tag_change = [this](const dwmipc::TagChangeEvent& ev) { this->on_tag_change(ev); };
+        m_ipc->subscribe(dwmipc::Event::TAG_CHANGE);
+        m_log.info("%s: Initialized tags", name());
+      }
+
+      if (m_layout_label) {
+        m_layouts = m_ipc->get_layouts();
+
+        // First layout is treated as default by dwm
+        m_default_layout = &m_layouts->at(0);
+        m_current_layout = find_layout(m_bar_mon->layout.address.cur);
+
+        if (m_layout_click) {
+          m_secondary_layout = find_layout(m_secondary_layout_symbol);
+
+          if (m_secondary_layout == nullptr) {
+            throw module_error("Secondary layout symbol does not exist");
+          }
+        }
+
+        // Initialize layout symbol
+        update_layout_label(m_bar_mon->layout.symbol.cur);
+
+        // This event is only needed to update the layout label
+        m_ipc->on_layout_change = [this](const dwmipc::LayoutChangeEvent& ev) { on_layout_change(ev); };
+        m_ipc->subscribe(dwmipc::Event::LAYOUT_CHANGE);
+        m_log.info("%s: Initialized layout label", name());
+      }
+
+      // These events are only necessary to update the focused window title
+      if (m_title_label) {
+        update_title_label();
+        m_ipc->on_client_focus_change = [this](const dwmipc::ClientFocusChangeEvent& ev) {
+          this->on_client_focus_change(ev);
+        };
+        m_ipc->subscribe(dwmipc::Event::CLIENT_FOCUS_CHANGE);
+
+        m_ipc->on_focused_title_change = [this](const dwmipc::FocusedTitleChangeEvent& ev) {
+          this->on_focused_title_change(ev);
+        };
+        m_ipc->subscribe(dwmipc::Event::FOCUSED_TITLE_CHANGE);
+        m_log.info("%s: Initialized title label", name());
+      }
+
+      if (m_floating_label) {
+        update_floating_label();
+        m_ipc->on_focused_state_change = [this](const dwmipc::FocusedStateChangeEvent& ev) {
+          this->on_focused_state_change(ev);
+        };
+        m_ipc->subscribe(dwmipc::Event::FOCUSED_STATE_CHANGE);
+        m_log.info("%s: Initialized floating label", name());
+      }
+
+      // This event is for keeping track of the currently focused monitor
+      m_ipc->on_monitor_focus_change = [this](const dwmipc::MonitorFocusChangeEvent& ev) {
+        this->on_monitor_focus_change(ev);
+      };
+      m_ipc->subscribe(dwmipc::Event::MONITOR_FOCUS_CHANGE);
+      m_log.info("%s: Subscribed to monitor focus change", name());
+    } catch (const dwmipc::IPCError& err) {
+      throw module_error(err.what());
+    }
+  }
+
+  void dwm_module::stop() {
```

Comment:
I'm not sure we need this function. The pointer is automatically reset and the underlying connection closed when the module is deconstructed. Also, it shouldn't be possible to catch an exception here because the destructor for the connection object (or any destructor for that matter) should never throw an exception.